sys/filedesc.c

```diff
@@ -170,6 +170,23 @@ int fdtab_install_file(fdtab_t *fdt, file_t *f, int *fd) {
   return 0;
 }
 
+int fdtab_install_file_at(fdtab_t *fdt, file_t *f, int fd) {
+  assert(f);
+  assert(fdt);
+
+  mtx_scoped_lock(&fdt->fdt_mtx);
+
+  if (fd < 0 || fd > fdt->fdt_nfiles)
+    return -EBADF;
+
+  if (fd_is_used(fdt, fd))
+    fd_free(fdt, fd);
+
+  fdt->fdt_files[fd] = f;
```

Comment:
I think you also need to `fd_mark_used`. In the original `fdtab_install_file` this was done by `fd_alloc`, but since you don't use it here, you need to manually mark that the file descriptor is now used.